src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AppendFileStreamStrategy.cs

```diff
@@ -0,0 +1,193 @@
+        {
+            if (value < Length)
+            {
+                throw new IOException(SR.IO_SetLengthAppendTruncate);
+            }
+
+            FileStreamHelpers.SetFileLength(_fileHandle, value);
+        }
+
+        public override IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback? callback, object? state)  => throw new NotSupportedException(SR.NotSupported_UnreadableStream);
+
+        public override int EndRead(IAsyncResult asyncResult)  => throw new NotSupportedException(SR.NotSupported_UnreadableStream);
+
+        public override int ReadByte() => throw new NotSupportedException(SR.NotSupported_UnreadableStream);
+
+        public override int Read(byte[] buffer, int offset, int count) => throw new NotSupportedException(SR.NotSupported_UnreadableStream);
+
+        public override int Read(Span<byte> buffer) => throw new NotSupportedException(SR.NotSupported_UnreadableStream);
+
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken) => throw new NotSupportedException(SR.NotSupported_UnreadableStream);
+
+        public override ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken = default) => throw new NotSupportedException(SR.NotSupported_UnreadableStream);
+
+        public override unsafe void WriteByte(byte value) =>
+            Write(new ReadOnlySpan<byte>(&value, 1));
+
+        public override void Write(byte[] buffer, int offset, int count) =>
+            Write(new ReadOnlySpan<byte>(buffer, offset, count));
+
+        public override void Write(ReadOnlySpan<byte> buffer)
+        {
+            if (_fileHandle.IsClosed)
+            {
+                ThrowHelper.ThrowObjectDisposedException_FileClosed();
+            }
+
+            int r = RandomAccess.WriteAtOffset(_fileHandle, buffer, WriteOffset);
+            Debug.Assert(r >= 0, $"RandomAccess.WriteAtOffset returned {r}.");
+        }
+
+        public override IAsyncResult BeginWrite(byte[] buffer, int offset, int count, AsyncCallback? callback, object? state) =>
+            TaskToApm.Begin(WriteAsync(buffer, offset, count), callback, state);
+
+        public override void EndWrite(IAsyncResult asyncResult) =>
+            TaskToApm.End(asyncResult);
+
+        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken) =>
+            WriteAsyncCore(new ReadOnlyMemory<byte>(buffer, offset, count), cancellationToken).AsTask();
+
+        public override ValueTask WriteAsync(ReadOnlyMemory<byte> source, CancellationToken cancellationToken) =>
+#pragma warning disable CA2012 // The analyzer doesn't know the internal AsValueTask is safe.
+            WriteAsyncCore(source, cancellationToken).AsValueTask();
+#pragma warning restore CA2012
+
+        private ValueTask<int> WriteAsyncCore(ReadOnlyMemory<byte> source, CancellationToken cancellationToken)
```

Comment:
Does RandomAccess.WriteAtOffsetAsync guarantee to write out everything it was given, or might it write less?  If the latter, I'm realizing now that both this and the UnixFileStreamStrategy are potentially broken, in that their WriteAsync methods could then end up doing only a partial write even though the stream contract guarantees everything will be written.